packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/OutputTypes/ChangeTrackingDictionaryProvider.cs

```diff
@@ -0,0 +1,404 @@
+        private readonly CSharpType _dictionary;
+        private readonly CSharpType _IDictionary;
+        private readonly CSharpType _IReadOnlyDictionary;
+        private readonly CSharpType _IEnumerator;
+        private readonly CSharpType _keyValuePair;
+        private readonly FieldDeclaration _innerDictionaryField;
+        private readonly DictionaryExpression _innerDictionary;
+        private readonly MethodSignature _ensureDictionarySignature;
+
+        private InvokeInstanceMethodExpression EnsureDictionary { get; init; }
+        private BoolExpression IsUndefined { get; } = new BoolExpression(new MemberExpression(This, "IsUndefined"));
+
+        private ChangeTrackingDictionaryProvider(SourceInputModel? sourceInputModel)
+            : base(sourceInputModel)
+        {
+            WhereClause = Where.NotNull(_tKey);
+            _indexParam = new Parameter("key", null, _tKey, null, ValidationType.None, null);
+            _IDictionary = new CSharpType(typeof(IDictionary<,>), _tKey, _tValue);
+            _dictionary = new CSharpType(typeof(Dictionary<,>), _tKey, _tValue);
+            _IReadOnlyDictionary = new CSharpType(typeof(IReadOnlyDictionary<,>), _tKey, _tValue);
+            _IEnumerator = new CSharpType(typeof(IEnumerator<>), new CSharpType(typeof(KeyValuePair<,>), _tKey, _tValue));
+            _keyValuePair = new CSharpType(typeof(KeyValuePair<,>), _tKey, _tValue);
+            _innerDictionaryField = new FieldDeclaration(FieldModifiers.Private, new CSharpType(typeof(IDictionary<,>), _tKey, _tValue), "_innerDictionary");
+            _innerDictionary = new DictionaryExpression(_tKey, _tValue, new VariableReference(_IDictionary, _innerDictionaryField.Declaration));
+            _ensureDictionarySignature = new MethodSignature("EnsureDictionary", null, null, MethodSignatureModifiers.Public, _IDictionary, null, Array.Empty<Parameter>());
+            EnsureDictionary = This.Invoke(_ensureDictionarySignature);
+        }
+
+        protected override TypeSignatureModifiers GetDeclarationModifiers()
+        {
+            return TypeSignatureModifiers.Internal;
+        }
+
+        public override string Name => "ChangeTrackingDictionary";
+
+        protected override CSharpType[] BuildTypeArguments()
+        {
+            return new[] { _tKey, _tValue };
+        }
+
+        protected override FieldDeclaration[] BuildFields()
+        {
+            return new[] { _innerDictionaryField };
+        }
+
+        protected override CSharpType[] BuildImplements()
+        {
+            return new[] { _IDictionary, _IReadOnlyDictionary };
+        }
+
+        protected override CSharpMethod[] BuildConstructors()
+        {
+            var constructors = new List<CSharpMethod>
+            {
+                DefaultConstructor(),
+                ConstructorWithDictionary(),
+                ConstructorWithReadOnlyDictionary()
+            };
+
+            return constructors.ToArray();
```

Comment:
can we just initialize the array directly instead of doing `toarray` since we know the elements it will contain?

Reply:
yes this makes sense!